### PR TITLE
Various improvements

### PR DESCRIPTION
**Pull request by amotl**
Dear Jan-Piet and Ben,

we have another round of improvements to *mqttwarn* ready. This is the foundation of the implementation of our prototypic Hiveeyes Schwarmalarm #193 on top of *mqttwarn*, which is the preliminary alerting component of the [Hiveeyes system]. We are building upon one of our former contributions, see also [Incorporate topic names into topic targets] and #169, #170.

The changes are mainly about overall robustness, especially when having more pressure on the job queue. The shutdown behavior of *mqttwarn* hasn't been optimal before, we often had cases where *mqttwarn* wouldn't stop at all on `SIGTERM` because the worker (`def processor`) already had been shut down while `def cleanup` was waiting for the notification job queue to drain, resulting in deadlock situations only to be mitigated by `kill -9` on our systems.

Along the lines, we added and improved some log messages here and there, integrated some minor features not worth a separate pull request in our opinion and also added a conveniency routine to monkeypatch the xmpppy library code for mitigating a known and popular problem when connecting to XMPP servers with TLS support (can be used optionally, thus non-invasive).

The details:
- Make “num_workers” setting configurable
- Add possibility to send log messages to stderr
- Improve reporting about log level
- Log warning message if transform_data cannot be formatted properly
- Improve worker robustness and shutdown behavior re. job queue draining
- Improve logging and robustness for service plugins smtp and xmpp
- Add monkey patch for xmpp service plugin to mitigate problem with SSLSocket X.509 issuer

Please let us know what you think about this. In the meanwhile, we will prepare the main PR mentioned above. Thanks in advance for your efforts!

Cheers, Andreas.

[Hiveeyes system]: https://hiveeyes.org/docs/system/
[Incorporate topic names into topic targets]: https://github.com/jpmens/mqttwarn/wiki/Incorporating-topic-names#incorporate-topic-names-into-topic-targets

### CODE REVIEW

File: services/smtp.py [hunk 1]
```diff
@@ -26,8 +27,13 @@ def plugin(srv, item):
     msg['From']         = sender
     msg['X-Mailer']     = srv.SCRIPTNAME
 
+    if not smtp_addresses:
+        srv.logging.warn("Skipped sending SMTP notification to %s, "
+                         "no addresses configured: %s" % (item.target, smtp_addresses))
```
Comment (sumnerboy12) [inline services/smtp.py]: Isn't the second parameter here redundant?

File: services/xmpp.py [hunk 2]
```diff
@@ -17,8 +17,13 @@ def plugin(srv, item):
     password = item.config['password']
     text = item.message
 
+    if not xmpp_addresses:
+        srv.logging.warn("Skipped sending XMPP notification to %s, "
+                         "no addresses configured: %s" % (item.target, xmpp_addresses))
```
Comment (sumnerboy12) [inline services/xmpp.py]: Same here?

**Comment by sumnerboy12** [comment]
Looks really good Andreas - thank you! I will leave JP to comment further but I took a good look around your project and am mightily impressed - I can imagine this will be a very useful reference project for others want to setup similar monitoring applications using MQTT/mqttwarn. Thanks again for sharing your code with our little project!

File: services/smtp.py (hunk 1 above)
Comment (jpmens) [inline services/smtp.py]: Yes, because it'll be `None`. I will remove after merge

**Comment by jpmens** [comment]
@amotl this looks excellent thank you. And thank you *particularly* for finally finding and fixing the suboptimal shutdown behavior. ;-)

**Comment by jpmens** [comment]
Merged!

File: services/smtp.py (hunk 1 above)
Comment (amotl) [inline services/smtp.py]: Sure, the second parameter will be None or an empty list. Thanks for spotting this!

File: services/xmpp.py (hunk 2 above)
Comment (amotl) [inline services/xmpp.py]: Right. :-)

**Comment by amotl** [comment]
Thanks!

**Pull request merged**